scvi/external/scar/_module.py

```diff
@@ -0,0 +1,460 @@
+            latent_distribution=latent_distribution,
+            deeply_inject_covariates=deeply_inject_covariates,
+            use_observed_lib_size=use_observed_lib_size,
+            **vae_kwargs,
+        )
+        self.sparsity = sparsity
+        self.ambient_profile = ambient_profile
+
+        cat_list = [n_batch] + list([] if n_cats_per_cov is None else n_cats_per_cov)
+
+        # decoder goes from n_latent-dimensional space to n_input-d data
+        n_input_decoder = n_latent + n_continuous_cov
+        self.decoder = DecoderSCAR(
+            n_input_decoder,
+            n_input,
+            n_cat_list=cat_list,
+            n_layers=n_layers,
+            n_hidden=n_hidden,
+            inject_covariates=deeply_inject_covariates,
+            scale_activation=scale_activation,
+            sparsity=self.sparsity,
+        )
+
+    @auto_move_data
+    def generative(
+        self,
+        z,
+        library,
+        batch_index,
+        cont_covs=None,
+        cat_covs=None,
+        size_factor=None,
+        y=None,
+        transform_batch=None,
+    ):
+        """Runs the generative model."""
+        # TODO: refactor forward function to not rely on y
+        # Likelihood distribution
+        if cont_covs is None:
+            decoder_input = z
+        elif z.dim() != cont_covs.dim():
+            decoder_input = torch.cat(
+                [z, cont_covs.unsqueeze(0).expand(z.size(0), -1, -1)], dim=-1
+            )
+        else:
+            decoder_input = torch.cat([z, cont_covs], dim=-1)
+
+        if cat_covs is not None:
+            categorical_input = torch.split(cat_covs, 1, dim=1)
+        else:
+            categorical_input = tuple()
+
+        if transform_batch is not None:
+            batch_index = torch.ones_like(batch_index) * transform_batch
+
+        if not self.use_size_factor_key:
+            size_factor = library
+
+        px_scale, px_noise_ratio, px_r, px_rate, px_dropout = self.decoder(
+            self.dispersion,
+            decoder_input,
+            size_factor,
+            batch_index,
+            *categorical_input,
+            y,
+        )
+        if self.dispersion == "gene-label":
+            px_r = F.linear(
+                one_hot(y, self.n_labels), self.px_r
+            )  # px_r gets transposed - last dimension is nb genes
+        elif self.dispersion == "gene-batch":
+            px_r = F.linear(one_hot(batch_index, self.n_batch), self.px_r)
+        elif self.dispersion == "gene":
+            px_r = self.px_r
+
+        # remove estimated noise
+        px_scale = px_scale * (1 - px_noise_ratio)
+        px_rate = px_rate * (1 - px_noise_ratio)
+        pamb_scale = self.ambient_profile.to(px_scale.device) * px_noise_ratio
+        pamb_rate = pamb_scale * torch.exp(size_factor)
+        px_r = torch.exp(px_r)
+
+        if self.gene_likelihood == "zinb":
+            px = ZeroInflatedNegativeBinomial(
+                mu=px_rate,
```

Comment:
I get it, posterior_predictive_sample works because of what you're doing here, pretty clever. I would add a comment somewhere about this because a priori it wouldn't make sense that `posterior_predictive_sample` removes background.

Reply:
I think I'll actually change this functionality and create a separate function to get the denoised counts. This function will have two options for retrieving denoised counts:

(background: scAR models both an estimated native and ambient count distribution)

1. Use the estimated native counts directly (as it is done in scAR)
2. Use the estimated ambient counts and remove them from the original counts (more conservative, this is how [cellbender is doing it](https://github.com/broadinstitute/CellBender/blob/cb2d209d5aedffe71e28947bc5b7859600aef64d/cellbender/remove_background/infer.py#L531))

Reply:
@ricomnl offered to implement it as well (and using jax perhaps so much faster), but of course this kind of contribution takes a significant amount of work so I can't promise any timeline.

Reply:
@cnk113 agreed! I guess the main difference between scAR and Cellbender is that scAR uses heuristics to define an ambient distribution and then learns to recreate it, while cellbender learns it de-novo from data. Cellbender also includes barcode swapping so the architecture is a bit more involved which is why I'd rather implement it separately. 
That being said, it would be quite simple to add the approach I mentioned above in case folks prefer a more conservative removal method in scAR

Reply:
I removed the cellbender flavor as the results seem a bit off (see comparison below) and as this is untested with this model I think I'd rather add this functionality with the cellbender model itself